File: Y2/CPE231/cpe231-algo-2025/src/main/java/th/ac/kmutt/cpe/algorithm/supawit/algorithms/searching/BinarySearch.java
package th.ac.kmutt.cpe.algorithm.supawit.searching;

import java.util.ArrayList;

public class BinarySearch<T extends Comparable<T>> implements SearchStrategy<T> {
  public int search(ArrayList<T> list, T target) {
    return searchRecursive(list, target, 0, list.size() - 1);
  }

  private int searchRecursive(ArrayList<T> list, T target, int left, int right) {
    if (left > right) {
      return -1;
    }
    int mid = left + (right - left) / 2;
    T midVal = list.get(mid);
    int cmp = midVal.compareTo(target);

    if (cmp == 0) {
      return mid;
    } else if (cmp < 0) {
      return searchRecursive(list, target, mid + 1, right);
    } else {
      return searchRecursive(list, target, left, mid - 1);
    }
  }
}

File: Y1/CPE111/02-sorting-searching/binary-search.c
#include <stdio.h>

int search(int *arr, int x, int l, int r) {
  while (l <= r) {
    int mid = l + (r - l) / 2;
    if (arr[mid] == x) {
      printf("Found %d at index %d.\n", x, mid);
      break;
    } else if (arr[mid] < x) {
      l = mid + 1;
    } else {
      r = mid - 1;
    }
  }

  if (l > r) {
    printf("Not found %d.\n", x);
  }
}

int main(int argc, char const *argv[]) {
  int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
  int n = sizeof(arr) / sizeof(arr[0]);
  int target = 5;

  int left = 0;
  int right = n - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;
    if (arr[mid] == target) {
      printf("Found %d at index %d.\n", target, mid);
      break;
    } else if (arr[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  if (left > right) {
    printf("Not found %d.\n", target);
  }

  return 0;
}
